R/renv.R

```diff
@@ -308,3 +308,56 @@ renv_switch_r_version <- function(version = NULL
 
   return(invisible(TRUE))
 }
+
+
+#' Downgrade an renv project to a specific RSPM snapshot
+#'
+#' @description
+#'   This functions aims to be used within a "snapshot-centered project
+#'   workflow" and can be used to downgrade all packages to an RSPM snapshot listed in
+#'   `renv.lock`.
+#'
+#'   While the main purpose is downgrade packages which exist in a higher
+#'   version, this function can also be used to restore a clean state of the
+#'   project library outside of a downgrade scenario.
+#'
+#'   Under the hood, it records all packages installed in the `renv` project
+#'   library and restores these with the RSPM snapshot found in Line 7 of
+#'   `renv.lock`.
+#' @return Called for its side-effect.
+#' @export
+#' @seealso renv_switch_r_version
+#'
+#' @examples
+#' \dontrun{
+#' renv_downgrade()
+#' }
+renv_downgrade <- function() {
+  requireNamespace("renv", quietly = TRUE)
+
+  # check if renv.lock exists
+  checkmate::assert_file("renv.lock", )
+
+  # to ensure repos is set correctly
+  source(".Rprofile")
+
+  # list all packages from renv library (should always be the renv lib
+  # if renv.lock exists)
+  installed_pkgs <- unname(utils::installed.packages(
+    lib.loc = .libPaths()[1]
+  )[, "Package"])
+
+  snapshot_date <- stringr::str_extract(
+    readLines("renv.lock", n = 7)[7],
+    "[0-9]{4}-[0-9]{2}-[0-9]{2}"
+  )
+  cli::cli_alert_info("Reinstalling all packages using RSPM snapshot
+    {.field {snapshot_date}}.", wrap = TRUE)
+
+  renv::install(installed_pkgs)
```

Comment:
What happens with GitHub packages here?

Reply:
Good point, need to check.

Reply:
In general, `renv::install()` is able to look up remote sources from `renv.lock` and match the package name.

There are two different scenarios though

## GH only pkg

It will install the exact same GH snapshot and only downgrade the CRAN ones.
Tested with {tic} which is a GH only pkg.

## Pkg on CRAN but GH version installed

For packages which are also on CRAN and `renv.lock` was using a GH build, `renv_downgrade()` downgrades this pkg to the respective CRAN snapshot.
This might not be desired but I don't see how this can be avoided. I'll add a dominant note in the help page.